api/src/main/java/com/capitalone/dashboard/webhook/github/GitHubCommitV3.java

```diff
@@ -0,0 +1,318 @@
+                Object userObject = restClient.getAsObject(authorObject, "user");
+                String authorLogin = (userObject == null) ? "unknown" : restClient.getString(userObject, "login");
+                commit.setScmAuthorLogin(authorLogin);
+
+                if (StringUtils.isEmpty(authorLDAPDN)) {
+                    String authorLDAPDNFetched = StringUtils.isEmpty(authorLogin) ? null : getLDAPDN(repoUrl, authorLogin, gitHubWebHookToken);
+                    commit.setScmAuthorLDAPDN(authorLDAPDNFetched);
+                }
+            }
+
+            int numberChanges = 0;
+            if (cObj.get("added") instanceof List) {
+                numberChanges += ((List) cObj.get("added")).size();
+            }
+            if (cObj.get("removed") instanceof List) {
+                numberChanges += ((List) cObj.get("removed")).size();
+            }
+            if (cObj.get("modified") instanceof List) {
+                numberChanges += ((List) cObj.get("modified")).size();
+            }
+
+            commit.setNumberOfChanges(numberChanges);
+
+            setCommitPullNumber(commit);
+
+            setCollectorItemId(commit, collectorItem);
+            commitsList.add(commit);
+        }
+
+        // For a merge commit for "Rebase and Merge"
+        setCommitPullNumbersForRebaseAndMergeCommit(commitsList);
+
+        return commitsList;
+    }
+
+    protected void setCommitPullNumbersForRebaseAndMergeCommit(List<Commit> commitsList) {
+
+        List<Commit> commitsWithPullNumber = Optional.ofNullable(commitsList)
+                                            .orElseGet(Collections::emptyList).stream()
+                                            .filter(commit -> !StringUtils.isEmpty(commit.getPullNumber()))
+                                            .collect(Collectors.toList());
+
+        // In case of Rebase and Merge, only the last commit in the list of commits on the "merge commit" json should have the PR number.
+        // This is because, there should be a corresponding PR in the DB whose "merge_commit_sha" matches the last commit in the list of commits
+        if (!CollectionUtils.isEmpty(commitsWithPullNumber)
+                && (commitsWithPullNumber.size() == 1)
+                && (commitsList.size() > 1)) {
+            Commit commitWithPR = commitsWithPullNumber.get(0);
+            commitsList.forEach(commit -> {commit.setPullNumber(commitWithPR.getPullNumber());});
+        }
+    }
+
+    protected void setCommitPullNumber (Commit commit) {
+        GitRequest pr = gitRequestRepository.findByScmRevisionNumberOrScmMergeEventRevisionNumber(commit.getScmRevisionNumber());
+        if (pr == null) {
+            pr = gitRequestRepository.findByCommitScmRevisionNumber(commit.getScmRevisionNumber());
+        }
+        if (pr != null) {
+            commit.setPullNumber(pr.getNumber());
+        }
+    }
+
+    protected void setCollectorItemId (Commit commit, CollectorItem collectorItem) {
```

Comment:
Can this method be refactored such that it can be reused by all 3 functionalities - commits, PRs and Issues?

Reply:
I see what you are saying, but since each of the setCollectorItemId methods, although they follow a similar pattern, use different object types and repository types. So I think even if I refactored them to take care of the three cases (commits, PRs and Issues), its not going to be pretty. The method would be one big monolithic method with a bunch of if-else conditions ..